src/reader/to_geos.rs

```diff
@@ -0,0 +1,277 @@
+    byte_order: Endianness,
+    num_coords: usize,
+    scratch: &mut Vec<f64>,
+) -> GResult<geos::CoordSeq> {
+    let (has_z, has_m, dim_size) = match dim {
+        WKBDimension::Xy => (false, false, 2),
+        WKBDimension::Xyz => (true, false, 3),
+        WKBDimension::Xym => (false, true, 3),
+        WKBDimension::Xyzm => (true, true, 4),
+    };
+    let num_ordinates = dim_size * num_coords;
+
+    // If the byte order matches native endianness, we can potentially use zero-copy
+    if byte_order == NATIVE_ENDIANNESS {
+        let ptr = unsafe { buf.as_ptr().add(offset as usize) };
+
+        // On platforms with unaligned memory access support, we can construct the coord seq
+        // directly from the raw parts without copying to the scratch buffer.
+        #[cfg(any(target_arch = "aarch64", target_arch = "x86_64"))]
+        {
+            let coords_f64 =
+                unsafe { &*core::ptr::slice_from_raw_parts(ptr as *const f64, num_ordinates) };
+            geos::CoordSeq::new_from_buffer(coords_f64, num_coords, has_z, has_m)
+        }
+
+        // On platforms without unaligned memory access support, we need to copy the data to the
+        // scratch buffer to make sure the data is aligned.
+        #[cfg(not(any(target_arch = "aarch64", target_arch = "x86_64")))]
+        {
+            unsafe {
+                scratch.clear();
+                scratch.reserve(num_ordinates);
+                scratch.set_len(num_ordinates);
+                std::ptr::copy_nonoverlapping(
+                    ptr,
+                    scratch.as_mut_ptr() as *mut u8,
+                    num_ordinates * std::mem::size_of::<f64>(),
+                );
+                geos::CoordSeq::new_from_buffer(scratch.as_slice(), num_coords, has_z, has_m)
+            }
+        }
+    } else {
+        // Need to convert byte order
+        match byte_order {
+            Endianness::BigEndian => {
+                save_f64_to_scratch::<BigEndian>(scratch, buf, offset, num_ordinates);
+            }
+            Endianness::LittleEndian => {
+                save_f64_to_scratch::<LittleEndian>(scratch, buf, offset, num_ordinates);
+            }
+        }
+        geos::CoordSeq::new_from_buffer(scratch.as_slice(), num_coords, has_z, has_m)
+    }
+}
+
+fn save_f64_to_scratch<B: ByteOrder>(
+    scratch: &mut Vec<f64>,
+    buf: &[u8],
+    offset: u64,
+    num_ordinates: usize,
+) {
+    scratch.clear();
+    scratch.reserve(num_ordinates);
+    // Safety: we have already reserved the capacity, so we can set the length safely.
+    // Justification: rewriting the loop to not use Vec::push makes it many times faster,
+    // since it eliminates several memory loads and stores for vector's length and capacity,
+    // and it enables the compiler to generate vectorized code.
+    #[allow(clippy::uninit_vec)]
+    unsafe {
+        scratch.set_len(num_ordinates);
+    }
+    assert!(offset + num_ordinates as u64 * 8 <= buf.len() as u64);
```

Comment:
This optimization is critical for parsing WKB with non-native endianness (big-endian). Please refer to https://www.notion.so/wherobots/Benchmarking-using-real-world-data-and-improving-performance-of-in-memory-join-23aba32ec027805ab7e4fae2d780f5de?source=copy_link#23bba32ec0278029b31ecfe250eef2d5 for detailed info about this optimization.